java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,239 @@
+import arcs.core.util.Time
+
+typealias ParticleConstructor = suspend () -> Particle
+typealias ParticleRegistration = Pair<ParticleIdentifier, ParticleConstructor>
 
 /**
  * Base helper class for [ArcHost] implementations to provide implementation of
  * registration.
+ *
+ * @property initialParticles The initial set of [Particle]s that this host contains.
  */
-abstract class AbstractArcHost(
-    private var particles: MutableList<ParticleIdentifier> = mutableListOf()
-) : ArcHost {
-    // TODO: fix this, qualifiedName is not supported on JS
+abstract class AbstractArcHost(vararg initialParticles: ParticleRegistration) : ArcHost {
+    private val particles: MutableMap<ParticleIdentifier, ParticleConstructor> = mutableMapOf()
+    private val runningArcs: MutableMap<String, ArcHostContext> = mutableMapOf()
+
+    init {
+        initialParticles.toList().associateByTo(particles, { it.first }, { it.second })
+    }
+
     override val hostId = this::class.className()
 
-    protected fun registerParticle(particle: ParticleIdentifier) {
-        particles.add(particle)
+    protected fun registerParticle(particle: ParticleIdentifier, constructor: ParticleConstructor) {
+        particles.put(particle, constructor)
     }
 
     protected fun unregisterParticle(particle: ParticleIdentifier) {
         particles.remove(particle)
     }
 
-    override suspend fun registeredParticles(): List<ParticleIdentifier> = particles
+    override suspend fun registeredParticles(): List<ParticleIdentifier> = particles.keys.toList()
+
+    // VisibleForTesting
+    protected fun getArcHostContext(arcId: String) = runningArcs[arcId]
+
+    /** Subclasses may override this to load persistent context state. */
+    protected suspend fun lookupOrCreateArcHostContext(partition: Plan.Partition): ArcHostContext =
+        runningArcs[partition.arcId] ?: ArcHostContext()
+
+    /** Subclasses may override this to store persistent context state. */
+    protected suspend fun updateArcHostContext(arcId: String, context: ArcHostContext) {
+        runningArcs[arcId] = context
+    }
 
     override suspend fun startArc(partition: Plan.Partition) {
-        // TODO: not implemented yet
+        val context = lookupOrCreateArcHostContext(partition)
+
+        // Arc is already currently running, don't restart it
+        if (context.arcState == ArcState.Running) {
+            return
+        }
+
+        // Can't restart a deleted arc
+        if (context.arcState == ArcState.Deleted) {
+            return
+        }
+
+        for (particleSpec in partition.particles) {
+            context.particles[particleSpec] = startParticle(particleSpec, context)
+        }
+
+        performLifecycleForContext(context)
+        maybeRequestResurrection(context)
+
+        context.arcState = ArcState.Running
+        updateArcHostContext(partition.arcId, context)
+    }
+
+    suspend fun startParticle(spec: Plan.Particle, context: ArcHostContext): ParticleContext {
```

Comment:
Please either make this private, or add copious KDoc. `startParticle` feels like a **super** important lifecycle step that should have some really detailed documentation for future engineers.

Actually, I think it's probably so important that even if it does become private, it should still have very detailed comments.